src/sprite_resource.c

```diff
@@ -0,0 +1,884 @@
+static void *SpriteResource_GetData(const SpriteResource *spriteRes);
+static void SpriteResource_FreeData(SpriteResource *spriteRes);
+static int SpriteResourceTableEntryNARC_GetEntryCount(const SpriteResourceTableEntryNARC *entries);
+
+SpriteResourceCollection *SpriteResourceCollection_New(int capacity, enum SpriteResourceType type, enum HeapId heapID)
+{
+    SpriteResourceCollection *spriteResources = Heap_AllocFromHeap(heapID, sizeof(SpriteResourceCollection));
+    spriteResources->collection = ResourceCollection_New(capacity, heapID);
+    spriteResources->resources = Heap_AllocFromHeap(heapID, sizeof(SpriteResource) * capacity);
+
+    memset(spriteResources->resources, 0, sizeof(SpriteResource) * capacity);
+
+    spriteResources->capacity = capacity;
+    spriteResources->count = 0;
+    spriteResources->type = type;
+
+    return spriteResources;
+}
+
+void SpriteResourceCollection_Delete(SpriteResourceCollection *spriteResources)
+{
+    GF_ASSERT(spriteResources);
+    GF_ASSERT(spriteResources->collection);
+    GF_ASSERT(spriteResources->resources);
+
+    SpriteResourceCollection_Clear(spriteResources);
+
+    ResourceCollection_Delete(spriteResources->collection);
+    spriteResources->collection = NULL;
+
+    Heap_FreeToHeap(spriteResources->resources);
+    spriteResources->resources = NULL;
+
+    Heap_FreeToHeap(spriteResources);
+    spriteResources = NULL;
+}
+
+SpriteResource *SpriteResourceCollection_AddFromTable(SpriteResourceCollection *spriteResources,
+    const SpriteResourceTable *table,
+    int index,
+    enum HeapId heapID)
+{
+    GF_ASSERT(spriteResources);
+    GF_ASSERT(table);
+    GF_ASSERT(table->count > index);
+    GF_ASSERT(spriteResources->type == table->type);
+
+    SpriteResource *spriteRes = SpriteResourceCollection_AllocResource(spriteResources);
+    GF_ASSERT(spriteRes);
+
+    if (table->isNARC == FALSE) {
+        SpriteResourceTableEntryFile *fileEntry = (SpriteResourceTableEntryFile *)table->entries + index;
+        GF_ASSERT(SpriteResourceCollection_IsIDUnused(spriteResources, fileEntry->id) == TRUE);
+        SpriteResourceCollection_InitResFromFile(
+            spriteResources,
+            spriteRes,
+            fileEntry->filename,
+            fileEntry->id,
+            fileEntry->vramType,
+            fileEntry->paletteIndex,
+            table->type,
+            heapID);
+    } else {
+        SpriteResourceTableEntryNARC *narcEntry = (SpriteResourceTableEntryNARC *)table->entries + index;
+        GF_ASSERT(SpriteResourceCollection_IsIDUnused(spriteResources, narcEntry->id) == TRUE);
+        SpriteResourceCollection_InitRes(
+            spriteResources,
+            spriteRes,
+            narcEntry->narcIndex,
+            narcEntry->memberIndex,
+            narcEntry->compressed,
+            narcEntry->id,
+            narcEntry->vramType,
+            narcEntry->paletteIndex,
+            table->type,
+            heapID,
+            0);
```

Comment:
**suggestion:** This call-pattern makes me question how meaningful `isNARC` is as a member-name. Maybe `loadFromNARC`?

Reply:
I chose `isNARC` because it indicates the type of `entries` (being either `SpriteResourceTableEntryFile` or `SpriteResourceTableEntryNARC`, but I don't mind changing it